src/image.c

```diff
@@ -1975,7 +1978,11 @@ GdipGetImagePaletteSize (GpImage *image, INT* size)
         if (image->active_bitmap->pixel_format == PixelFormat4bppIndexed)
                 palette_entries = 16;
 
-	*size = (palette_entries - 1) * sizeof(ARGB) + sizeof(ColorPalette);
+	if (palette_entries == 0)
+		*size = sizeof(ColorPalette);
+	else
+		*size = (palette_entries - 1) * sizeof(ARGB) + sizeof(ColorPalette);
+
```

Comment:
This is the "fix" that caused the tests to fail.

Previously, we'd actually subtract(!) from the total size of ColorPalette, leading to potentially under allocating memory.

Reply:
Yeah I agree this is a good fix, the question is just what the best way of coordinating the fix with updating the unit tests in Mono is.

Reply:
we do have a collection of FIXME's already that require changes to the Mono unit tests right? maybe we should batch them together and enable them all in one go with the next release?

Reply:
Hmm, actually if you look at the comment in GdipGetImagePalette it seems to explicitly mention and contradict the fix:

```
 		/* ColorPalette definition includes one ARGB member (ARGB Entries[1]). In reality there are 
 		 * Count entries allocated, where Count can be 0 (and required to substract the ARGB size)
 		 */
 		/* coverity[buffer_alloc] */
 		image->active_bitmap->palette = (ColorPalette*) GdipAlloc (sizeof(ColorPalette) - sizeof(ARGB));
```

It allocates `-1 * sizeof(ARGB) + sizeof(ColorPalette)` there!

Reply:
Okay so I did a bit of digging in to this. The fact is that in the following code, `GdipGetImagePaletteSize` returns `sizeof(ColorPalette)` if there is the current count is 0.  [Wine also has this check](https://github.com/wine-mirror/wine/blob/c579a81a6353dd7c843270561cd5763ff895de40/dlls/gdiplus/image.c#L2291-L2294).

```
static void test_getImagePaletteSize ()
{
	GpStatus status;
	GpImage *bitmapImage = getImage ("test.bmp");
	INT size;

	status = GdipGetImagePaletteSize (bitmapImage, &size);
	assertEqualInt (status, Ok);
	assertEqualInt (sizeof(ColorPalette), 12);
	assertEqualInt (size, 12);

	GdipDisposeImage (bitmapImage);
}
```

However, I was pretty much wrong on all fronts in GdipGetImagePalette. I've added a bunch of tests in a new commit. All of these failed previously with my change, and also with my fix...

Take a look and let me know